src/neighborsearch/filtered.jl

```diff
@@ -0,0 +1,189 @@
+# ------------------------------------------------------------------
+# Licensed under the MIT License. See LICENSE in the project root.
+# ------------------------------------------------------------------
+
+"""
+    FilteredSearch(method, nmax=Inf; maxpercategory, maxpersector, metric=Euclidean())
+
+A method for searching at most `nmax` neighbors using `method`. Extra
+constraints available: `maxpercategory` and `maxpersector`. The neighbors are
+sorted using `metric` distance (priority to the nearest). To preserve the order
+of the initial neighbors returned by `method`, set `metric = nothing`.
+
+## Max per category
+
+It can be a `NamedTuple` or `Dict`, where the first element is the property name
+and the second defines the max neighbor per available categories of the property.
+In the example below, no more than 2 neighbors of same geometry are selected.
+
+`FilteredSearch(method, maxpercategory = (geometry = 2,))`
+```ascii
+     _________
+  //     . □   \\           ○  Search point
+ //      . ▩    \\        ▩ ▲  Neighbors selected
+// △ ▲   . ▩     \\       □ △  Neighbors ignored
+ ........○........
+\\ △ ▲   .       //
+ \\      .      //
+  \\ △   . □   //
+     ‾‾‾‾‾‾‾‾‾
+```
+
+## Max per sector
+
+Only allow a max number of neighbors inside each sector. The sectors are the
+cartesian quadrants (2-D) or octants (3-D). If the neighborhood is an `Ellipsoid`,
+the sectors will match the rotated quadrants/octants.
+
+`FilteredSearch(method, maxpersector = 2)`
+```ascii
+     _________
+  //     . □   \\           ○  Search point
+ //      . ▩    \\        ▩ ▲  Neighbors selected
+// ▲ ▲   . ▩     \\       □ △  Neighbors ignored
+ ........○........
+\\ ▲ ▲   .       //
+ \\      .      //
+  \\ △   . ▩   //
+     ‾‾‾‾‾‾‾‾‾
+```
+"""
+struct FilteredSearch{M<:NeighborSearchMethod} <: BoundedNeighborSearchMethod
+  method::M
+  nmax
+  maxpercategory
+  maxpersector
+  metric
+end
+
+FilteredSearch(method::M, nmax=Inf; maxpercategory=nothing,
+               maxpersector=nothing, metric=Euclidean()) where {M} =
```

Comment:
It would make more sense if the metric was the one of the method if there is one

Reply:
That'd probably be more intuitive indeed.
Maybe we could let as default `metric=:same` or something similar